### DISCUSSION TITLE
differentiate controlled and uncontrolled NumberInput

### DISCUSSION THREAD

**Discussion by vaynevayne** [mantinedev/mantine#3781] [Q&A] [answered]
> [NumberInput](https://mantine.dev/core/number-input/) component types for value, defaultValue and onChange props were changed. It now expects value to be number | '' (number or empty string) instead of number | undefined. This change was made to address multiple bugs that happened because it was not possible to differentiate controlled and uncontrolled NumberInput.

https://mantine.dev/changelog/6-0-0/#numberinput-breaking-changes

Can you simply distinguish between controlled and uncontrolled like this? I guess adding numberInput to the string type will result in some run-time errors
```jsx
type ControlledProps<T> = {
 value: T
 onChange: (value: T) => void
  defaultValue?: never
}

type UncontrolledProps<T> = {
  value?: never
  onChange?: never
  defaultValue: T
}

type NumberInputProps = {
[index:string]:any
} & (ControlledProps<number | undefined> | UncontrolledProps<number| undefined>)
```

**Comment by rtivital** [comment, marked as answer]
> Can you simply distinguish between controlled and uncontrolled like this?

Yes, controlled input value is not undefined. Uncontrolled input value is undefined.
